lib/web_ui/lib/src/engine/text/canvas_paragraph.dart

```diff
@@ -240,6 +240,32 @@ class CanvasParagraph implements ui.Paragraph {
     return lines.map((ParagraphLine line) => line.lineMetrics).toList();
   }
 
+  @override
+  EngineLineMetrics? getLineMetricsAt(int lineNumber) {
+    return 0 <= lineNumber && lineNumber < lines.length
+      ? lines[lineNumber].lineMetrics
+      : null;
+  }
+
+  @override
+  int get numberOfLines => lines.length;
+
+  @override
+  int? getLineNumberAt(int codeUnitOffset) => _findLine(codeUnitOffset, 0, lines.length);
+
+  int? _findLine(int codeUnitOffset, int startLine, int endLine) {
+    if (endLine <= startLine || codeUnitOffset < lines[startLine].startIndex || lines[endLine - 1].endIndex <= codeUnitOffset) {
+      return null;
+    }
+    if (endLine == startLine + 1) {
+      return startLine;
+    }
+    // endLine >= startLine + 2 thus we have
+    // startLine + 1 <= midIndex <= endLine - 1
+    final int midIndex = (startLine + endLine) ~/ 2;
+    return _findLine(codeUnitOffset, midIndex, endLine) ?? _findLine(codeUnitOffset, startLine, midIndex);
+  }
```

Comment:
There's code here that does the same thing:
https://github.com/flutter/engine/blob/85e1dd9b79b64a007056ad8ddd3e7c7239d69709/lib/web_ui/lib/src/engine/text/canvas_paragraph.dart#L226-L232

Yours seems to be more performant (because binary search), so it would be nice to use it in `getLineBoundary()`.